src/rust/src/lib.rs

```diff
@@ -105,39 +138,52 @@ fn model_matrix(data: List) -> Result<Robj> {
     Ok(robj)
 }
 
-fn process_integer_column(
-    column: &Robj,
-    col_name: &str,
-    output_column_names: &mut Vec<String>,
-    output: &mut [f64],
-) {
-    output_column_names.push(col_name.to_string());
-    zip(column.as_integer_slice().unwrap().iter(), output.iter_mut()).for_each(
-        |(integer_element, output)| {
-            if integer_element.is_na() {
-                *output = f64::na();
-            } else {
-                *output = *integer_element as f64;
+enum Calculation<'b> {
+    FactorColumn {
+        column: Robj,
+        nrow: usize,
+        output: &'b mut [f64],
+    },
+    IntegerColumn {
+        column: Robj,
+        output: &'b mut [f64],
+    },
+    DoubleColumn {
+        column: Robj,
+        output: &'b mut [f64],
+    },
+    LogicalColumn {
+        column: Robj,
+        output: &'b mut [f64],
+    },
+}
+unsafe impl<'a> Send for Calculation<'a> {}
```

Comment:
Should these be gated behind `if cfg!(feature = "rayon")`?

Reply:
It can be gated, but it does not need to.

Reply:
It could be useful if you wanted to do parallelisation without using `rayon`, but instead spawn your own threads, or however you like.